src/server/pkg/storage/fileset/writer.go

```diff
@@ -141,76 +103,57 @@ func (w *Writer) nextIdx(idx *index.Index) error {
 
 func (w *Writer) checkPath(p string) error {
 	if w.idx != nil {
-		if w.idx.Path > p {
-			return errors.Errorf("can't write path (%s) after (%s)", p, w.idx.Path)
+		if w.idx.Path == p {
+			return errors.Errorf("cannot write same path (%s) twice", p)
 		}
-		parent := parentOf(p)
-		if w.idx.Path < parent {
-			return errors.Errorf("cannot write path (%s) without first writing parent (%s)", p, parent)
+		if w.idx.Path > p {
+			return errors.Errorf("cannot write path (%s) after (%s)", p, w.idx.Path)
 		}
 	}
 	return nil
 }
 
 // Delete creates a delete operation for a file.
-func (w *Writer) Delete(p string) error {
+// TODO: Check path order.
+func (w *Writer) Delete(p string, tags ...string) error {
 	idx := &index.Index{
 		Path: p,
-		FileOp: &index.FileOp{
-			Op: index.Op_DELETE,
-		},
+		File: &index.File{},
+	}
+	for _, tag := range tags {
+		idx.File.Parts = append(idx.File.Parts, &index.Part{Tag: tag})
 	}
-	return w.nextIdx(idx)
+	return w.deletive.WriteIndex(idx)
 }
 
 // Copy copies a file to the file set writer.
 func (w *Writer) Copy(file File) error {
 	idx := file.Index()
-	if idx.FileOp.DataRefs != nil {
-		return w.copyIndex(idx)
-	}
-	switch idx.FileOp.Op {
-	case index.Op_APPEND:
-		return w.Append(idx.Path, func(fw *FileWriter) error {
-			hdr, err := file.Header()
-			if err != nil {
-				return err
-			}
-			return WithTarFileWriter(fw, hdr, func(tfw *TarFileWriter) error {
-				return tfw.Copy(getContentDataOps(idx.FileOp.DataOps))
-			})
-		})
-	case index.Op_OVERWRITE:
-		return w.Overwrite(idx.Path, func(fw *FileWriter) error {
-			hdr, err := file.Header()
-			if err != nil {
-				return err
-			}
-			return WithTarFileWriter(fw, hdr, func(tfw *TarFileWriter) error {
-				return tfw.Copy(getContentDataOps(idx.FileOp.DataOps))
-			})
-		})
-	case index.Op_DELETE:
-		return w.Delete(idx.Path)
-	}
-	return nil
-}
-
-func (w *Writer) copyIndex(idx *index.Index) error {
 	copyIdx := &index.Index{
 		Path: idx.Path,
-		FileOp: &index.FileOp{
-			Op:      idx.FileOp.Op,
-			DataOps: idx.FileOp.DataOps,
+		File: &index.File{
+			Parts: idx.File.Parts,
 		},
 		SizeBytes: idx.SizeBytes,
 	}
 	if err := w.nextIdx(copyIdx); err != nil {
 		return err
 	}
-	for _, dataRef := range idx.FileOp.DataRefs {
-		if err := w.cw.Copy(dataRef); err != nil {
-			return err
+	// Copy the file data refs if they are resolved.
+	if idx.File.DataRefs != nil {
```

Comment:
See my comment at the beginning. I think we should use the empty tag for this and just store it as parts.

Reply:
Same as above. I disagree, but this is likely to change so it does not really matter.